crud/stats/init.lua

```diff
@@ -256,6 +257,79 @@ function stats.wrap(func, op, opts)
     end
 end
 
+local storage_stats_schema = { tuples_fetched = 'number', tuples_lookup = 'number' }
+--- Callback to collect storage tuples stats (select/pairs).
+--
+-- @function update_fetch_stats
+--
+-- @local
+--
+-- @tab storage_stats
+--  Statistics from select storage call.
+--
+-- @number storage_stats.tuples_fetched
+--  Count of tuples fetched during storage call.
+--
+-- @number storage_stats.tuples_lookup
+--  Count of tuples looked up on storages while collecting response.
+--
+-- @string space_name
+--  Name of space.
+--
+-- @treturn boolean Returns `true`.
+--
+local function update_fetch_stats(storage_stats, space_name)
+    dev_checks(storage_stats_schema, 'string')
+
+    if not stats.is_enabled() then
+        return true
+    end
+
+    registry.observe_fetch(
+        storage_stats.tuples_fetched,
+        storage_stats.tuples_lookup,
+        space_name
+    )
+
+    return true
+end
+
+--- Returns callback to collect storage tuples stats (select/pairs).
+--
+-- @function get_fetch_callback
+--
+-- @treturn[1] function `update_fetch_stats` function to collect tuples stats.
+-- @treturn[2] function Dummy function, if stats disabled.
+--
+function stats.get_fetch_callback()
```

Comment:
TBH, I don't get why we can call `stats.update_map_reduces()`, but can't call `update_fetch_stats()` in the same way. IOW, why we need the `status.get_fetch_callback()` function, but not just `stats.update_fetch()`?

Reply:
Map reduce counter is updated once per request. Tuple counters are updated once per replicaset for all requests aside of select and pairs, where it is chunks count (at min one per storage replicaset). And we save one if here. Hm.

If this is was the main point, I would not bother so much. I hope it is the job of a tracing jit compiler.

Reply:
Found the following comment in `select_old.lua`:

```lua
-- Stats extracted with callback here and not passed
-- outside to wrapper because fetch for pairs can be
-- called even after pairs() return from generators.
```

However it does not makes things clear for me (aside of that it is not written here). It seems, the idea is offer some kind of protection against `crud.cfg()` while an iterator is open.

What is correct: it keeps writting / don't writting metrics despite `crud.cfg()` calls.

What is strange: it doesn't keep the old registry. So, if we changed it from 'local' to 'metrics' or vice versa, part of the tuples will be counted in the old registry and part of them will be counted in the new one.

I found this behaviour okay, to be honest. I call it strange, because we can achieve (almost) the same without the callback.

Well, let's just look on it once again and discuss.

Reply:
At least, if I undestood the code correctly, we'll not try to write into a destroyed registry. Neither now, nor if we'll call the functon directly.

Reply:
As far as I remember, the main point of this decision was to not introduce any additional load if stats is not required. But now it really seems overcomplicating. Reworked

Reply:
Thanks! Closing.